### PR TITLE
Email validations for sign up and sign in to avoid SQL Injection flag

### PR DESCRIPTION
**Pull request by domducsutcliffe**

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -692,4 +692,4 @@ RUBY VERSION
    ruby 2.7.6p219
 
 BUNDLED WITH
-   2.1.4
+  2.1.4
```
Comment (tim-s-ccs) [inline Gemfile.lock:695]: Looks like you've got an extra space here

Comment (tim-s-ccs) [inline Gemfile.lock:695]: Looks like you've got an extra space here

File: features/facilities_management/rm3830/cognito/validations/sign_in_validation.feature [hunk 2]
```diff
@@ -12,6 +12,23 @@ Feature: Sign in to my account - RM3830 - Validations
...
+    Then I should see the following error messages:
+      | You must provide your email address in the correct format, like name@example.com  |
+
+    Examples:
+      | email   |  
+      | local@  |
+      | @domain |
+      | @       |
+
+      #works
```
Comment (tim-s-ccs) [inline features/facilities_management/rm3830/cognito/validations/sign_in_validation.feature:29]: This comment is unnecessary and you should have one blank lines between examples

File: features/facilities_management/rm3830/cognito/validations/sign_up_user_validations.feature [hunk 3]
```diff
@@ -20,7 +20,20 @@ Feature: Sign up user - RM3830 - Validations
     Examples:
       | email         | error_message                                                       |
       |               | Enter an email address in the correct format, like name@example.com |
-      | Test@test.com | Email address cannot contain any capital letters                    |
+      | Test@test.com | Email address cannot contain any capital letters                    |   
+
+
+Scenario Outline: Incorrect email format on sign up & off safelist
```
Comment (tim-s-ccs) [inline features/facilities_management/rm3830/cognito/validations/sign_up_user_validations.feature:26]: What is this test doing that the one on line 38 isn't? Are you really testing that the email is in the correct format?

File: app/services/cognito/sign_up_user.rb [hunk 4]
```diff
@@ -2,7 +2,8 @@ module Cognito
   class SignUpUser < BaseService
     include ActiveModel::Validations
 
-    validates :email, presence: true, format: { with: /\A[^A-Z]*\z/ }
+    validates :email, presence: true, format: { with: /\A[^A-Z]*\z/, message: :invalid } # validates lowercase
+    validates_format_of :email, with: /\A[^\s^@]+@[^\s^@]+\z/, message: :blank # validates x@y
```
Comment (tim-s-ccs) [inline app/services/cognito/sign_up_user.rb:6]: These comments are unnecessary

File: app/services/cognito/sign_in_user.rb [hunk 5]
```diff
@@ -5,6 +5,8 @@ class SignInUser < BaseService
     attr_accessor :error, :needs_password_reset, :needs_confirmation
 
     validates_presence_of :email, :password
+    validates_format_of :email, with: /\A[^A-Z]*\z/, message: :invalid
```
Comment (tim-s-ccs) [inline app/services/cognito/sign_in_user.rb:8]: You have added validations to make sure the email does not have any capital letters but there are no tests for it (hint: there is a reason we don't need to validate that the email is all lowercase)

**Review by tim-s-ccs — CHANGES_REQUESTED** [review]
Great job Dom 😃, just a few things for you to look at

File: Gemfile.lock (hunk 1 above)
Comment (domducsutcliffe) [reply Gemfile.lock:695]: Yep fixed

File: app/services/cognito/sign_up_user.rb [hunk 6]
```diff
@@ -2,8 +2,9 @@ module Cognito
   class SignUpUser < BaseService
     include ActiveModel::Validations
 
-    validates :email, presence: true, format: { with: /\A[^A-Z]*\z/ }
+    validates :email, presence: true, format: { with: /\A[^A-Z]*\z/, message: :invalid }
```
Comment (tim-s-ccs) [inline app/services/cognito/sign_up_user.rb:5]: You did not need to change this line and the email

File: app/services/cognito/sign_up_user.rb [hunk 7]
```diff
@@ -2,8 +2,9 @@ module Cognito
...
     validates :password, presence: true, length: { within: 8..200 }
+    validates_format_of :email, with: /\A[^\s^@]+@[^\s^@]+\z/, message: :blank
```
Comment (tim-s-ccs) [inline app/services/cognito/sign_up_user.rb:7]: Move this line up by one so it is with the other email validation

**Review by tim-s-ccs — CHANGES_REQUESTED** [review]
Have a look at the files you've changed and think about what changes should and should not be there https://github.com/Crown-Commercial-Service/crown-marketplace/pull/3410/files

File: app/models/facilities_management/rm3830/procurement_contact_detail.rb [hunk 8]
```diff
@@ -3,7 +3,7 @@ module RM3830
     class ProcurementContactDetail < ApplicationRecord
       validates :name, presence: true, length: { maximum: 50 }, on: %i[new_invoicing_contact_details new_authorised_representative new_notices_contact_details]
       validates :job_title, presence: true, length: { maximum: 150 }, on: %i[new_invoicing_contact_details new_authorised_representative new_notices_contact_details]
-      validates :email, presence: true, format: { with: /\A([\w+\-].?)+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i }, on: %i[new_invoicing_contact_details new_authorised_representative new_notices_contact_details]
+      validates :email, presence: true, format: { with: /\A([\w+-].?)+@[a-z\d-]+(\.[a-z]+)*\.[a-z]+\z/i }, on: %i[new_invoicing_contact_details new_authorised_representative new_notices_contact_details]
```
Comment (tim-s-ccs) [inline app/models/facilities_management/rm3830/procurement_contact_detail.rb:6]: Why have you made changes here?

File: app/services/cognito/forgot_password.rb [hunk 9]
```diff
@@ -3,7 +3,7 @@ class ForgotPassword < BaseService
     include ActiveModel::Validations
     attr_reader :email, :error
 
-    validates :email, format: { with: /\A([\w+\-].?)+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i }
+    validates :email, format: { with: /\A([\w+-].?)+@[a-z\d-]+(\.[a-z]+)*\.[a-z]+\z/i }
```
Comment (tim-s-ccs) [inline app/services/cognito/forgot_password.rb:6]: Why have you made changes here?

**Review by tim-s-ccs — CHANGES_REQUESTED** [review]
Just a couple of other unexpected changes, if you are not sure why they are there you can ask me

**Review by tim-s-ccs — APPROVED** [review]
LGTM 👍 Now, you just need to do it legacy too

**Pull request merged**